### PR TITLE
[GDARV-258] - I3 batt switch rev3

### PR DESCRIPTION
**Pull request by AubreyC**
[GDARV-258](https://planckaero.atlassian.net/browse/GDARV-258)

Changes:
- support rev2: `BATT_MONITOR=16` and rev3 `BATT_MONITOR=17`
- `BATT_DISCONNECT`: enable / disable the battery disconnect logic 
- Battery disconnect logic: 
    - send `BATT_DISCO_EN` based on armed status (never allow disco when armed) && `MCU_alive`
    - `BATT_KILL`: not armed && tether power is off && `MCU_alive`

### CODE REVIEW

**Review by CraftyJax — APPROVED** [review]

File: ArduCopter/events.cpp [hunk 1]
```diff
@@ -315,6 +315,21 @@ void Copter::failsafe_tether_on_event()
     }
 }
 
+
+void Copter::battery_switch_status_check()
+{
+    for(uint8_t i = 0; i < battery.num_instances(); i++) {
+        if (battery.get_type(i) == AP_BattMonitor_Params::BattMonitor_Type::BattMonitor_TYPE_ANALOG_VOLTAGE_AND_CURRENT_AND_GPIO_REV3) {
+
+            bool batt_disco_en = !motors->armed() && AP::battery().mcu_alive();
```
Comment (503e) [inline ArduCopter/events.cpp]: missing function parameter. Should be `i`.

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+private:
+  AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev;
+  bool _is_using_battery = false;
+  bool _mcu_alive = false;
+
+  struct send_state {
+    bool batt_disco_en = false;
+    bool batt_kill = false;
+  } _send_state;
+  bool _send_required = false;
```
Comment (503e) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.h:53]: Do `_send_state` and `_send_required` need to be thread safe? Is the timer running on its own thread?

**Review by 503e — CHANGES_REQUESTED** [review]
I can see an argument for making a `AP_BattMonitor_Analog_GPIO` base class, but I think the approach is good for now. Just a couple comments that could use a review to make sure they are okay.

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.h (hunk 2 above)
Comment (AubreyC) [reply libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.h:53]: Good question, I thought it was, but looking more at the docs it seems that it actually is not the case (at least not all the time / all platforms), for reference I found similar logic here: 
https://github.com/planckaero/ardupilot/blob/master/libraries/AP_Notify/OreoLED_I2C.cpp#L305-L326

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp [hunk 3]
```diff
@@ -84,6 +86,10 @@ void AP_BattMonitor_Analog_GPIO_rev3::timer() {
 
 void AP_BattMonitor_Analog_GPIO_rev3::set_batt_disco_en(bool enable)
 {
+  if(!_batt_disconnect_enable)
+    return;
+
+  WITH_SEMAPHORE(_sem);
   if(_params._disconnect_enable > 0 && _send_state.batt_disco_en !=  enable) {
```
Comment (503e) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp]: Remove `_params._disconnect_enable > 0` from condition. Now redundant.

**Review by 503e — APPROVED** [review]
Probably should be tested again.

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  }
+}
+
+void AP_BattMonitor_Analog_GPIO_rev3::set_batt_kill(bool enable)
+{
+  if(!_batt_disconnect_enable)
+    return;
+
+  WITH_SEMAPHORE(_sem);
+  if(_send_state.batt_kill !=  enable) {
```
Comment (ptran) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp]: Extra space

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    }
+}
+
+void AP_BattMonitor_Analog_GPIO_rev3::set_batt_disco_en(bool enable)
+{
+  if(!_batt_disconnect_enable)
+    return;
+
+  WITH_SEMAPHORE(_sem);
+  if(_send_state.batt_disco_en !=  enable) {
```
Comment (ptran) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp]: Extra space

File: libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+      if (new_mcu_alive) {
+          gcs().send_text(MAV_SEVERITY_CRITICAL, "Battery MCU alive");
+      } else {
+          gcs().send_text(MAV_SEVERITY_CRITICAL, "Battery MCU not alive");
+      }
+  }
+  _mcu_alive = new_mcu_alive;
+
+  // AP_BATTMONITOR_FET_EN_TETHER_REV3: Active high
+  bool new_is_using_battery = (bool)((buf & AP_BATTMONITOR_FET_EN_TETHER_REV3) == 0);
```
Comment (ptran) [inline libraries/AP_BattMonitor/AP_BattMonitor_Analog_GPIO_rev3.cpp]: This explicit cast is needed?

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 7]
```diff
@@ -545,6 +553,40 @@ bool AP_BattMonitor::reset_remaining(uint16_t battery_mask, float percentage)
     return ret;
 }
 
+void AP_BattMonitor::set_batt_disco_en(bool enable, uint8_t instance)
+{
+    if (instance < _num_instances && drivers[instance] != nullptr)
```
Comment (ptran) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:558]: Noticed some `if` statements are `if(` and some are `if (`

**Review by ptran — APPROVED** [review]
LGTM. Only left comments about formatting but not really a big deal.

**Pull request merged**